src/array/cuda/csr_mm.cu

```diff
@@ -70,30 +70,110 @@ std::pair<CSRMatrix, NDArray> CusparseSpgemm(
   CUSPARSE_CALL(cusparseSpGEMM_workEstimation(
       thr_entry->cusparse_handle, transA, transB, &alpha, matA, matB, &beta,
-      matC, dtype, CUSPARSE_SPGEMM_DEFAULT, spgemmDesc, &workspace_size1,
+      matC, dtype, alg, spgemmDesc, &workspace_size1,
       NULL));
   void* workspace1 = (device->AllocWorkspace(ctx, workspace_size1));
   // inspect the matrices A and B to understand the memory requiremnent
-  // for the next step
-  CUSPARSE_CALL(cusparseSpGEMM_workEstimation(
-      thr_entry->cusparse_handle, transA, transB, &alpha, matA, matB, &beta,
-      matC, dtype, CUSPARSE_SPGEMM_DEFAULT, spgemmDesc, &workspace_size1,
-      workspace1));
+  cusparseStatus_t e =
+    cusparseSpGEMM_workEstimation(thr_entry->cusparse_handle, transA,
+                                  transB, &alpha, matA, matB, &beta,
+                                  matC, dtype, alg, spgemmDesc,
+                                  &workspace_size1, workspace1);
+  // CUSPARSE_SPGEMM_DEFAULT not support getting num_prods > 2^31 -1
+  // and throws insufficient memory error within workEstimation call
+  if (e == CUSPARSE_STATUS_INSUFFICIENT_RESOURCES) {
+    // fall back to ALG2 to estimate num_prods
+    alg = CUSPARSE_SPGEMM_ALG2;
+    device->FreeWorkspace(ctx, workspace1);
+    // rerun cusparseSpGEMM_workEstimation
+    CUSPARSE_CALL(cusparseSpGEMM_workEstimation(thr_entry->cusparse_handle,
+                                                transA, transB, &alpha, matA,
+                                                matB, &beta, matC, dtype, alg,
+                                                spgemmDesc, &workspace_size1,
+                                                NULL));
+    void* workspace1 = (device->AllocWorkspace(ctx, workspace_size1));
+    CUSPARSE_CALL(cusparseSpGEMM_workEstimation(thr_entry->cusparse_handle,
+                                                transA, transB, &alpha, matA,
+                                                matB, &beta, matC, dtype, alg,
+                                                spgemmDesc, &workspace_size1,
+                                                workspace1));
+  } else {
+    CHECK(e == CUSPARSE_STATUS_SUCCESS) << "CUSPARSE ERROR: " << e;
+  }
+
+  // you can print num_prods, for better memory estimation
+  int64_t num_prods;
+  CUSPARSE_CALL(cusparseSpGEMM_getNumProducts(spgemmDesc, &num_prods));
+
+  // assume free GPU mem at least ~15G for below heuristics to work
+  // user-defined medium problem size (below will use DEFAULT)
+  int64_t MEDIUM_NUM_PRODUCTS = 400*1000*1000;
+  // user-defined large problem size (above will use ALG3)
+  int64_t LARGE_NUM_PRODUCTS  = 800*1000*1000;
+
+  // switch to ALG2/ALG3 for medium & large problem size
+  if (alg == CUSPARSE_SPGEMM_DEFAULT && num_prods > MEDIUM_NUM_PRODUCTS) {
+    // use ALG3 for very large problem
+    if (num_prods <= LARGE_NUM_PRODUCTS)
+      alg = CUSPARSE_SPGEMM_ALG2;
+    else
+      alg = CUSPARSE_SPGEMM_ALG3;
```

Comment:
The condition operator `? :` looks better.

Reply:
Updated.